<a href="https://colab.research.google.com/github/Ihzaamirunnadif/UTS_4112322007_ML/blob/main/UTS_4112322007_Machine_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Ihza Amirunnadif

NIM  : 4112322007


## import library

In [30]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.graph_objects as go

from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split, GridSearchCV, RepeatedStratifiedKFold
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC
from sklearn.metrics import (accuracy_score, precision_score,
                             recall_score, f1_score, classification_report,
                             confusion_matrix, roc_curve, auc)
from imblearn.over_sampling import SMOTE

sns.set_style("whitegrid")

Pada bagian ini, dilakukan proses import berbagai pustaka (library) yang dibutuhkan untuk mendukung proses pemrosesan data, pembuatan model, evaluasi model, hingga visualisasi hasil. Library seperti numpy dan pandas digunakan untuk pengolahan data numerik dan manipulasi dataset, sedangkan matplotlib, seaborn, dan plotly digunakan untuk visualisasi data. Untuk keperluan pra-pemrosesan data, digunakan StandardScaler dan MinMaxScaler dari pustaka sklearn.preprocessing yang berfungsi untuk menormalkan fitur agar memiliki skala yang seragam.

Selanjutnya, dari pustaka sklearn.model_selection diimpor beberapa fungsi seperti train_test_split untuk membagi data menjadi data latih dan data uji, serta GridSearchCV dan RepeatedStratifiedKFold untuk proses pencarian parameter terbaik. klasifikasi seperti Decision Tree, Gradient Boosting, K-Nearest Neighbors, Multi-Layer Perceptron, dan Support Vector Machine yang diimpor dari pustaka sklearn. Untuk mengevaluasi kinerja model, digunakan berbagai metrik seperti akurasi, presisi, recall, f1-score, dan area under curve (AUC) dari sklearn.metrics. Selain itu, digunakan pula teknik oversampling bernama SMOTE (Synthetic Minority Over-sampling Technique) dari pustaka imblearn untuk menangani ketidakseimbangan kelas dalam dataset. Terakhir, sns.set_style("whitegrid") digunakan untuk mengatur tampilan visualisasi agar lebih rapi dan mudah dibaca.

## **1. Eksplorasi Data**

### Memuat Dataset

In [31]:
df = pd.read_csv('https://raw.githubusercontent.com/Ihzaamirunnadif/UTS_4112322007_ML/refs/heads/main/data_training.csv')
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0              7.3             0.740         0.08             1.7      0.094   
1              8.1             0.575         0.22             2.1      0.077   
2             10.1             0.430         0.40             2.6      0.092   
3             12.9             0.500         0.55             2.8      0.072   
4              8.4             0.360         0.32             2.2      0.081   
..             ...               ...          ...             ...        ...   
852            6.7             1.040         0.08             2.3      0.067   
853            8.0             0.390         0.30             1.9      0.074   
854            7.4             0.350         0.33             2.4      0.068   
855            7.9             0.570         0.31             2.0      0.079   
856            7.5             0.520         0.40             2.2      0.060   

     free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                   10.0                  45.0  0.99576  3.24       0.50   
1                   12.0                  65.0  0.99670  3.29       0.51   
2                   13.0                  52.0  0.99834  3.22       0.64   
3                    7.0                  24.0  1.00012  3.09       0.68   
4                   32.0                  79.0  0.99640  3.30       0.72   
..                   ...                   ...      ...   ...        ...   
852                 19.0                  32.0  0.99648  3.52       0.57   
853                 32.0                  84.0  0.99717  3.39       0.61   
854                  9.0                  26.0  0.99470  3.36       0.60   
855                 10.0                  79.0  0.99677  3.29       0.69   
856                 12.0                  20.0  0.99474  3.26       0.64   

     alcohol  quality    Id  
0        9.8        5  1366  
1        9.2        5   103  
2       10.0        7   942  
3       10.9        6   811  
4       11.0        6   918  
..       ...      ...   ...  
852     11.0        4  1467  
853      9.0        5  1533  
854     11.9        6  1580  
855      9.5        6  1216  
856     11.8        6  1575  

[857 rows x 13 columns]

pada tahap ini, dataset kualitas anggur dimuat. hasilnya ditampilkan dalam bentuk dataframe yang memiliki 857 baris dan 13 kolom.

### Identifikasi Data

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857 entries, 0 to 856
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         857 non-null    float64
 1   volatile acidity      857 non-null    float64
 2   citric acid           857 non-null    float64
 3   residual sugar        857 non-null    float64
 4   chlorides             857 non-null    float64
 5   free sulfur dioxide   857 non-null    float64
 6   total sulfur dioxide  857 non-null    float64
 7   density               857 non-null    float64
 8   pH                    857 non-null    float64
 9   sulphates             857 non-null    float64
 10  alcohol               857 non-null    float64
 11  quality               857 non-null    int64  
 12  Id                    857 non-null    int64  
dtypes: float64(11), int64(2)
memory usage: 87.2 KB


Dataset `ckd-dataset-v2.csv` berisi data mengenai pasien yang diduga mengalami Chronic Kidney Disease (CKD), dengan total 200 entri dan 27 kolom. Berikut adalah deskripsi setiap kolom:

1. `bp (Diastolic)` (*int64*) - Tekanan darah diastolik pasien, dikodekan sebagai 0 (normal) dan 1 (tinggi).  
2. `bp limit` (*int64*) - Indikator apakah tekanan darah melewati batas normal.  
3. `sg` (*object*) - Nilai berat jenis urin dalam bentuk interval (misalnya, "1.009 - 1.011").  
4. `al` (*object*) - Tingkat albumin dalam urin, bisa berupa nilai seperti `< 0`, `≥ 4`, dll.  
5. `class` (*object*) - Kategori status CKD: `ckd` (terkena) atau `notckd` (tidak terkena).  
6. `rbc` (*int64*) - Sel darah merah dalam urin: 0 (normal), 1 (tidak normal).  
7. `su` (*object*) - Kadar gula dalam urin, dalam format teks seperti `< 0`.  
8. `pc` (*int64*) - Jumlah sel nanah (pus cells) dalam urin: 0 (tidak ada), 1 (ada).  
9. `pcc` (*int64*) - Gumpalan sel nanah: 0 (tidak ada), 1 (ada).  
10. `ba` (*int64*) - Kehadiran bakteri dalam urin: 0 (tidak), 1 (ada).  
11. `bgr` (*object*) - Kadar glukosa dalam darah (bentuk rentang nilai sebagai teks).  
12. `bu` (*object*) - Kadar urea dalam darah, dicatat dalam format teks numerik.  
13. `sod` (*object*) - Kadar natrium dalam darah, dalam bentuk teks angka.  
14. `sc` (*object*) - Kadar kreatinin dalam serum, nilai berbentuk string angka.  
15. `pot` (*object*) - Kadar kalium dalam darah.  
16. `hemo` (*object*) - Tingkat hemoglobin pasien.  
17. `pcv` (*object*) - Volume sel darah yang dipadatkan, dalam bentuk teks numerik.  
18. `rbcc` (*object*) - Jumlah sel darah merah, juga dicatat sebagai teks numerik.  
19. `wbcc` (*object*) - Jumlah sel darah putih dalam darah pasien.  
20. `htn` (*int64*) - Status hipertensi: 0 (tidak), 1 (ya).  
21. `dm` (*int64*) - Status diabetes: 0 (tidak), 1 (ya).  
22. `cad` (*int64*) - Riwayat penyakit jantung koroner: 0 (tidak), 1 (ya).  
23. `appet` (*int64*) - Nafsu makan: 0 (baik), 1 (buruk).  
24. `pe` (*int64*) - Pembengkakan (edema): 0 (tidak), 1 (ya).  
25. `ane` (*int64*) - Anemia: 0 (tidak), 1 (ya).  
26. `grf` (*object*) - Laju filtrasi glomerulus (GFR), dalam interval nilai sebagai teks.  
27. `age` (*object*) - Kelompok usia pasien, misalnya `< 12`, `12 - 20`, dst.

melihat informasi struktur dataset, dataset terdiri atas 857 baris dan 13 kolom, dengan masing-masing kolom memiliki tipe data

### Identifikasi Missing Value

In [33]:
print(df.isnull().sum())

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
Id                      0
dtype: int64


 proses identifikasi missing value atau nilai yang hilang dalam dataset menggunakan fungsi df.isnull().sum().Hasil yang ditampilkan memperlihatkan bahwa tidak ada satupun nilai yang hilang di seluruh kolom, karena semua kolom memiliki nilai nol (0) pada hasilnya.

In [34]:
df = df.drop(columns=['Id'])
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0              7.3             0.740         0.08             1.7      0.094   
1              8.1             0.575         0.22             2.1      0.077   
2             10.1             0.430         0.40             2.6      0.092   
3             12.9             0.500         0.55             2.8      0.072   
4              8.4             0.360         0.32             2.2      0.081   
..             ...               ...          ...             ...        ...   
852            6.7             1.040         0.08             2.3      0.067   
853            8.0             0.390         0.30             1.9      0.074   
854            7.4             0.350         0.33             2.4      0.068   
855            7.9             0.570         0.31             2.0      0.079   
856            7.5             0.520         0.40             2.2      0.060   

     free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                   10.0                  45.0  0.99576  3.24       0.50   
1                   12.0                  65.0  0.99670  3.29       0.51   
2                   13.0                  52.0  0.99834  3.22       0.64   
3                    7.0                  24.0  1.00012  3.09       0.68   
4                   32.0                  79.0  0.99640  3.30       0.72   
..                   ...                   ...      ...   ...        ...   
852                 19.0                  32.0  0.99648  3.52       0.57   
853                 32.0                  84.0  0.99717  3.39       0.61   
854                  9.0                  26.0  0.99470  3.36       0.60   
855                 10.0                  79.0  0.99677  3.29       0.69   
856                 12.0                  20.0  0.99474  3.26       0.64   

     alcohol  quality  
0        9.8        5  
1        9.2        5  
2       10.0        7  
3       10.9        6  
4       11.0        6  
..       ...      ...  
852     11.0        4  
853      9.0        5  
854     11.9        6  
855      9.5        6  
856     11.8        6  

[857 rows x 12 columns]

proses penghapusan kolom 'Id' dari dataset menggunakan perintah df.drop(columns=['Id']), dikarenakan  tidak memiliki pengaruh terhadap analisis atau pemodelan.Setelah kolom dihapus, dataset kini terdiri dari 12 kolom dan 857 baris.

## **2. Pra-Pemrosesan Data**

### Data Training & Data Testing

In [35]:
df["quality"].unique()

array([5, 7, 6, 4, 8, 3])

 proses pra-pemrosesan data khususnya untuk kolom "quality" yang menjadi target atau label dalam dataset. Perintah df["quality"].unique() digunakan untuk melihat nilai-nilai unik dalam kolom "quality", yang menunjukkan bahwa terdapat enam kelas berbeda yaitu 3, 4, 5, 6, 7, dan 8. Ini mengindikasikan bahwa kualitas wine dalam dataset dinilai dalam rentang tersebut, dengan nilai yang lebih tinggi menunjukkan kualitas yang lebih baik.

In [36]:
print(df.quality.value_counts())

quality
5    362
6    341
7    109
4     26
8     13
3      6
Name: count, dtype: int64


hasil dari df.quality.value_counts() menampilkan jumlah data untuk masing-masing nilai kualitas. Diketahui bahwa kelas kualitas 5 dan 6 mendominasi dataset dengan jumlah masing-masing 362 dan 341. Sedangkan kualitas 3, 4, 7, dan 8 memiliki jumlah data yang jauh lebih sedikit, terutama kelas 3 dan 8 yang hanya terdiri dari masing-masing 6 dan 13.

In [37]:
df["quality"].value_counts(normalize = True)

scaler = MinMaxScaler()

min_max_df = scaler.fit_transform(df.drop("quality", axis = 1))
min_max_df = pd.DataFrame(min_max_df, columns = df.columns[:-1])
X = min_max_df
y = df['quality']

oversample = SMOTE()

X, y = oversample.fit_resample(X, y)
X.shape, y.shape
y.value_counts()

quality
5    362
7    362
6    362
4    362
8    362
3    362
Name: count, dtype: int64

proses normalisasi data dan penyeimbangan kelas menggunakan teknik SMOTE (Synthetic Minority Over-sampling Technique) sebelum melakukan pemisahan data latih dan data uji. Data selain kolom target (quality) dinormalisasi menggunakan MinMaxScaler agar semua fitur berada dalam rentang 0 hingga 1, sehingga model tidak berat sebelah terhadap fitur tertentu karena perbedaan skala. Kemudian, dilakukan pemisahan variabel independen (X) dan dependen (y) dengan X berisi semua fitur dan y adalah label kualitas wine.

In [51]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

data dibagi menjadi data latih dan uji  menggunakan train_test_split, memastikan distribusi data yang adil untuk pelatihan dan evaluasi model.

## **3. Pemrosesan Data**

In [62]:
models = {
    'Decision Tree': DecisionTreeClassifier(random_state=4),
    'KNN': KNeighborsClassifier(),
    'SVM': SVC(),
    "GradBoost": GradientBoostingClassifier(),
    'MLP Neural Network': MLPClassifier(max_iter=1000),
}

evaluation_results = []

for model_name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    # metrik evaluasi
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')

    # hasil
    evaluation_results.append({
        'Model': model_name,
        'Accuracy': round(accuracy, 4),
        'Precision': round(precision, 4),
        'Recall': round(recall, 4),
        'F1 Score': round(f1, 4)
    })

/usr/local/lib/python3.11/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning:

Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.



dilakukan evaluasi terhadap beberapa algoritma klasifikasi, yaitu Decision Tree, K-Nearest Neighbors (KNN), Support Vector Machine (SVM), Gradient Boosting, dan MLP Neural Network. Masing-masing model dilatih menggunakan data latih (X_train dan y_train) dan diuji menggunakan data uji (X_test). Proses evaluasi dilakukan dengan menghitung empat metrik performa utama, yaitu akurasi, presisi, recall, dan F1-score dengan pendekatan rata-rata berbobot (average='weighted'). Hasil evaluasi dari tiap model disimpan dalam bentuk list of dictionary, yang berisi nama model dan nilai metrik evaluasinya.

## **4. Evaluasi Model**

### Matrik Evaluasi Model

In [63]:
# DataFrame
results_df = pd.DataFrame(evaluation_results)

fig = go.Figure(data=[go.Table(
    header=dict(
        values=[f"<b>{col}</b>" for col in results_df.columns],
        fill_color='white',
        line_color='black',
        align='center',
        font=dict(color='black', size=12)
    ),
    cells=dict(
        values=[results_df[col] for col in results_df.columns],
        fill_color='white',
        line_color='black',
        align='center',
        font=dict(color='black', size=12)
    )
)])

fig.update_layout(
    title=dict(
        text="Perbandingan Evaluasi Model",
        x=0.5,
        xanchor='center',
        font=dict(color='black', size=20)
    ),
    margin=dict(t=30)
)

fig.show()

 hasil visualisasi evaluasi performa lima model klasifikasi dalam bentuk tabel. Metrik yang ditampilkan mencakup akurasi, presisi, recall, dan F1-score. Berdasarkan tabel, model Gradient Boosting menunjukkan performa terbaik dengan nilai akurasi 0.8103 dan F1-score tertinggi sebesar 0.8056, yang mengindikasikan bahwa model ini paling konsisten dalam mengklasifikasikan data dengan benar. Sebaliknya, model SVM memiliki performa terendah di semua metrik, menunjukkan bahwa model ini kurang sesuai untuk dataset yang digunakan.

In [64]:
classifier = GradientBoostingClassifier()
classifier.fit(X_train, y_train)

GradientBoostingClassifier()

proses pelatihan model Gradient Boosting Classifier, Model diinisialisasi menggunakan GradientBoostingClassifier() dari pustaka sklearn.ensemble dan kemudian dilatih dengan data pelatihan X_train dan y_train melalui perintah .fit().

In [65]:
y_pred=classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.to_numpy().reshape(len(y_test),1)),1))
final_classifier = classifier

[[7 7]
 [6 5]
 [4 4]
 ...
 [4 5]
 [6 5]
 [3 3]]


proses prediksi dan pencocokan hasil, Model yang telah dilatih digunakan untuk memprediksi label dari data uji X_test dengan menggunakan metode .predict(). Hasil prediksi y_pred kemudian dibandingkan dengan label aktual y_test. Hasil dari keduanya ditampilkan berdampingan dengan menggunakan np.concatenate, yang membantu dalam memverifikasi apakah prediksi yang dilakukan sudah sesuai dengan label sebenarnya.

In [68]:
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[119   0   0   0   0   0]
 [  0  99   6   2   0   0]
 [  2  11  68  31   3   1]
 [  1   4  33  60  21   1]
 [  0   0   4  14 110  11]
 [  0   0   0   0   0 116]]


evaluasi performa model menggunakan confusion matrix. Matriks ini dihitung dengan fungsi confusion_matrix(y_test, y_pred) dari sklearn.metrics, yang menampilkan hasil klasifikasi untuk masing-masing kelas dalam bentuk tabel.

In [70]:
fin_test_data = pd.read_csv('https://raw.githubusercontent.com/Ihzaamirunnadif/UTS_4112322007_ML/refs/heads/main/data_testing.csv')
test_data = fin_test_data.drop('Id',axis=1)
test_data = sc.fit_transform(test_data)

masukan data testing dari github yang didefinisikan sebagai fin_test_data

In [71]:
final_pred = final_classifier.predict(test_data)
final_pred

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning:

X does not have valid feature names, but GradientBoostingClassifier was fitted with feature names



array([5, 5, 5, 6, 5, 7, 5, 7, 6, 5, 7, 6, 5, 6, 6, 5, 6, 6, 6, 6, 6, 5,
       6, 7, 5, 6, 6, 7, 5, 7, 5, 7, 6, 6, 5, 6, 5, 6, 7, 7, 7, 6, 7, 6,
       5, 5, 5, 7, 6, 5, 6, 5, 6, 6, 6, 7, 5, 6, 7, 5, 6, 6, 6, 7, 6, 6,
       6, 6, 6, 5, 7, 6, 6, 5, 5, 5, 6, 6, 6, 6, 6, 5, 6, 5, 6, 6, 5, 7,
       4, 4, 5, 6, 6, 6, 5, 6, 6, 7, 5, 6, 7, 6, 5, 6, 6, 7, 6, 7, 6, 7,
       6, 5, 6, 6, 5, 5, 4, 5, 5, 5, 7, 5, 5, 6, 5, 6, 6, 5, 4, 6, 7, 5,
       6, 6, 6, 5, 5, 7, 5, 7, 6, 5, 7, 6, 5, 5, 5, 6, 6, 6, 7, 6, 7, 6,
       7, 6, 6, 7, 6, 4, 5, 6, 7, 7, 7, 6, 7, 6, 4, 6, 6, 6, 6, 6, 6, 6,
       7, 6, 6, 5, 7, 7, 6, 6, 7, 5, 7, 5, 7, 6, 7, 6, 5, 6, 5, 6, 5, 6,
       6, 6, 4, 5, 6, 6, 7, 7, 7, 5, 7, 7, 5, 7, 5, 6, 6, 7, 5, 6, 5, 6,
       7, 5, 5, 7, 5, 6, 5, 6, 6, 5, 6, 7, 5, 6, 7, 7, 5, 7, 6, 5, 7, 7,
       7, 5, 4, 6, 6, 6, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5, 6, 6, 6, 6, 4,
       5, 5, 7, 7, 6, 6, 5, 6, 7, 7, 5, 5, 5, 5, 5, 5, 6, 6, 5, 5, 7, 6])

output diatas adalah hasil prediksi dari dataset testing yang dimasukan

In [72]:
sample_submit = pd.read_csv('https://raw.githubusercontent.com/Ihzaamirunnadif/UTS_4112322007_ML/refs/heads/main/hasilprediksi_3digitNIMterakhir.csv')
sample_submit

Id;Quality
0         222;
1        1514;
2         417;
3         754;
4         516;
..         ...
281      1147;
282       296;
283       170;
284      1439;
285       946;

[286 rows x 1 columns]

masukan file submition dari github yang didefinisikan sebagai sample_submit, kemudian save dengan kode dibawah ini:

In [73]:
output = pd.DataFrame({'Id':fin_test_data.Id,'Quality':final_pred})
output.to_csv('hasilprediksi_007.csv', index=False)
filename = "hasilprediksi_007.csv"
print("Your submission was successfully saved!")

Your submission was successfully saved!
